src/Common/EventNotifier.h

```diff
@@ -0,0 +1,92 @@
+#pragma once
+
+#include <vector>
+#include <mutex>
+#include <functional>
+#include <set>
+#include <map>
+#include <memory>
+#include <utility>
+#include <iostream>
+
+#include <base/types.h>
+#include <Common/HashTable/Hash.h>
+
+namespace DB
+{
+
+class EventNotifier
```

Comment:
Got it. One possible difficulty - we will have to recreate this watch each time an error happens and a new `zookeeper` is created. But I hope this logic will be enclosed in scope of MergeTreeRestarting thread.

Reply:
Yep, it will be quite natural to add this callback when starting up replication in RestartingThread and remove it when shutting it down. We can even use `replica_path` as "callback identified" and put this callback into `ZooKeeper::watches` (but in this case we will need to store extra flag in `ZooKeeper::watches` to distinguish it from real watches on removal). Or we can prepend something to `replica_path` to guarantee that no real watches with the same id exist (real paths start with '/')